### PR TITLE
[torch distributed] Implementing reduce_scatter_base

### PR DESCRIPTION
**Pull request by liangluofb**
Summary: Support flattened reduce_scatter.

Test Plan:
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/torch/lib/c10d:ProcessGroupNCCLTest
buck test mode/opt -c fbcode.enable_gpu_sections=true //caffe2/test/distributed:c10d

Differential Revision: D27876281

### CODE REVIEW

File: torch/lib/c10d/ProcessGroup.hpp [hunk 1]
```diff
@@ -257,7 +258,17 @@ class ProcessGroup : public torch::CustomClassHolder {
   virtual c10::intrusive_ptr<ProcessGroup::Work> reduce_scatter(
       std::vector<at::Tensor>& outputTensors,
       std::vector<std::vector<at::Tensor>>& inputTensors,
-      const ReduceScatterOptions& opts = ReduceScatterOptions()) = 0;
+      const ReduceScatterOptions& = ReduceScatterOptions()) = 0;
```
Comment (zhaojuanmao) [inline torch/lib/c10d/ProcessGroup.hpp:261]: nit: keep the same as before

File: torch/lib/c10d/ProcessGroup.hpp [hunk 2]
```diff
@@ -257,7 +258,17 @@ class ProcessGroup : public torch::CustomClassHolder {
...
+
+  virtual c10::intrusive_ptr<ProcessGroup::Work> _reduce_scatter_base(
+      at::Tensor&,
+      at::Tensor&,
+      const ReduceScatterOptions& opts = ReduceScatterOptions())
+      {
+        // does not support it by default
+        throw std::runtime_error("ProcessGroup does not support alltoall");
```
Comment (zhaojuanmao) [inline torch/lib/c10d/ProcessGroup.hpp:269]: nit: "ProcessGroup does not support _reduce_scatter_base"

File: test/distributed/test_c10d_nccl.py [hunk 3]
```diff
@@ -500,6 +500,25 @@ def reduce_scatter(outputs, input_lists, op):
             # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
             self.assertEqualIgnoreType(expected, output[i])
 
+    @requires_nccl()
+    def test_reduce_scatter_base_ops(self):
+        store = c10d.FileStore(self.file.name, self.world_size)
+        pg = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def reduce_scatter_base(output_t, input_t):
+            work = pg._allgather_base(output_t, input_t)
```
Comment (zhaojuanmao) [inline test/distributed/test_c10d_nccl.py:509]: it is testing _allgather_base, not _reduce_scatter_base?

also, would you please add a  test for dist._reduce_scatter_base() API as well? and the error cases

**Review by zhaojuanmao — COMMENTED** [review]
lgtm overall, some minor comments

File: torch/lib/c10d/ProcessGroup.hpp (hunk 1 above)
Comment (Luo-Liang) [reply torch/lib/c10d/ProcessGroup.hpp:261]: the linter issued a warning. I will overwrite that.

File: torch/lib/c10d/ProcessGroup.hpp (hunk 2 above)
Comment (Luo-Liang) [reply torch/lib/c10d/ProcessGroup.hpp:269]: Thanks!

File: test/distributed/test_c10d_nccl.py (hunk 3 above)
Comment (Luo-Liang) [reply test/distributed/test_c10d_nccl.py:509]: sg!

Comment (zhaojuanmao) [reply test/distributed/test_c10d_nccl.py:591]: why is '_allgather_base' called here?

File: torch/utils/_cpp_extension_versioner.py [hunk 4]
```diff
@@ -41,13 +41,13 @@ def bump_version_if_changed(self,
...
                                 is_python_module,
                                 is_standalone):
-        hash_value = 0
-        hash_value = hash_source_files(hash_value, source_files)
-        hash_value = hash_build_arguments(hash_value, build_arguments)
-        hash_value = update_hash(hash_value, build_directory)
-        hash_value = update_hash(hash_value, with_cuda)
-        hash_value = update_hash(hash_value, is_python_module)
-        hash_value = update_hash(hash_value, is_standalone)
+        # hash_value = 0
```
Comment (zhaojuanmao) [inline torch/utils/_cpp_extension_versioner.py:44]: why is this changed?

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -1253,67 +1253,67 @@ def _jit_compile(name,
...
-    elif verbose:
-        print('No modifications detected for re-loaded extension '
-              f'module {name}, skipping build step...')
-
-    if verbose:
-        print(f'Loading extension module {name}...')
-
-    if is_standalone:
-        return _get_exec_path(name, build_directory)
+    # if is_python_module and is_standalone:
```
Comment (zhaojuanmao) [inline torch/utils/cpp_extension.py:1256]: same here, why is this changed?

**Comment by zhaojuanmao** [comment]
codes lgtm, will wait until tests are green

**Review by zhaojuanmao — APPROVED** [review]
would you please rebase and make sure tests are green before landing? thanks!
